From 77615b55353865222794500487ae0b4b70888efd Mon Sep 17 00:00:00 2001
From: Tim Snyder <snyder.tim@gmail.com>
Date: Tue, 29 Mar 2022 22:33:27 +0000
Subject: [PATCH 2/5] allow % commands in main specfile

---
 gcc/gcc.cc | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/gcc/gcc.cc b/gcc/gcc.cc
index 728332b81..c04b23d56 100644
--- a/gcc/gcc.cc
+++ b/gcc/gcc.cc
@@ -2388,7 +2388,8 @@ read_specs (const char *filename, bool main_p, bool user_p)
       /* Is this a special command that starts with '%'? */
       /* Don't allow this for the main specs file, since it would
 	 encourage people to overwrite it.  */
-      if (*p == '%' && !main_p)
+      /* ::conda-forge:: allow use of commands in main specs */
+      if (*p == '%')
 	{
 	  p1 = p;
 	  while (*p && *p != '\n')
