--- libgomp/secure_getenv.h	2021-07-14 23:18:04.405198104 +0000
+++ libgomp/secure_getenv.h	2021-07-14 23:18:44.306179642 +0000
@@ -25,7 +25,7 @@
 #define _SECURE_GETENV_H 1
 
 /* Secure getenv() which returns NULL if running as SUID/SGID.  */
-#ifndef HAVE_SECURE_GETENV
+#if __x86_64__
 #if defined (HAVE_UNISTD_H) && defined (HAVE_GETUID) \
   && defined (HAVE_GETEUID) && defined (HAVE_GETGID) \
   && defined (HAVE_GETEGID)
